File: 1706-where-will-the-ball-fall/1706-where-will-the-ball-fall.java
class Solution {
    public int[] findBall(int[][] grid) {
        int n = grid.length, m = grid[0].length, res[] = new int[m];
        for(int ball = 0; ball < m; ball++){
            int j1 = ball, j2;
            for(int i = 0; i < n; i++){
                j2 = j1 + grid[i][j1];
                if(j2 < 0 || j2 >= m || grid[i][j1] != grid[i][j2]) {
                    j1 = -1;
                    break;
                }
                j1 = j2;
            }
            res[ball] = j1;
        }
        return res;
    }
}

File: 1706-where-will-the-ball-fall/1706-where-will-the-ball-fall.cpp
class Solution {
public:
    int dfs(vector<vector<int>> &grid, int i, int j){
        if(i == grid.size()){
            return j;
        }
        
        // left to right
        if(grid[i][j] == 1 && j+1 < grid[0].size() && grid[i][j+1] == 1){
            return dfs(grid, i+1, j+1);
        } 
        // right to left
        else if(grid[i][j] == -1 && j-1 >= 0 && grid[i][j-1] == -1){
            return dfs(grid, i+1, j-1);
        } else {
            return -1;
        }
    }
    
    vector<int> findBall(vector<vector<int>>& grid) {
        int row = grid.size();
        int col = grid[0].size();
        vector<int> res(col);
        for(int j = 0; j < col; j++){
            res[j] = dfs(grid, 0, j);
        }
        return res;
    }
};